### PR TITLE
8335709: C2: assert(!loop->is_member(get_loop(useblock))) failed: must be outside loop

### PR DESCRIPTION
**Pull request by rwestrel**
When a `NeverBranch` is added to an infinite loop, the `NeverBanch`
and its projection that branches in the loop are assigned to the wrong
loop. The confuses the logic that tries to sink nodes out of loop.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8335709](https://bugs.openjdk.org/browse/JDK-8335709): C2: assert(!loop-&gt;is_member(get_loop(useblock))) failed: must be outside loop (**Bug** - P3)

### Reviewers
 * [Tobias Hartmann](https://openjdk.org/census#thartmann) (@TobiHartmann - **Reviewer**) 🔄 Re-review required (review applies to [114add0f](https://git.openjdk.org/jdk/pull/20231/files/114add0f221d0437331df69e82f8895bc11e6f28))
 * [Emanuel Peter](https://openjdk.org/census#epeter) (@eme64 - **Reviewer**)

### Contributors
 * Emanuel Peter `<epeter@openjdk.org>`

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/20231/head:pull/20231` \
`$ git checkout pull/20231`

Update a local copy of the PR: \
`$ git checkout pull/20231` \
`$ git pull https://git.openjdk.org/jdk.git pull/20231/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 20231`

View PR using the GUI difftool: \
`$ git pr show -t 20231`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/20231.diff">https://git.openjdk.org/jdk/pull/20231.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/20231#issuecomment-2236082065)

### CODE REVIEW

File: test/hotspot/jtreg/compiler/loopopts/InfiniteLoopBadControlNeverBranch.java [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+    public static void main(String[] args) throws InterruptedException {
+        Thread thread = new Thread(() -> test());
+        thread.setDaemon(true);
+        thread.start();
+        Thread.sleep(Utils.adjustTimeout(4000));
+    }
+
+    static void test() {
+        int i = 0;
+        while(true) {
```
Comment (TobiHartmann) [inline test/hotspot/jtreg/compiler/loopopts/InfiniteLoopBadControlNeverBranch.java:49]:
```suggestion
        while (true) {
```

**Review by TobiHartmann — APPROVED** [review]
That looks good to me. I submitted some testing and will report back once it passed.

**Review by eme64 — APPROVED** [review]
Thanks for the work and the credit :blush: 
LGTM!

**Comment by TobiHartmann** [comment]
All tests passed.

**Comment by rwestrel** [comment]
@TobiHartmann @eme64 thanks for the reviews

**Pull request closed without merging**